app/models/package_manager/base/bulk_version_updater.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+module PackageManager
+  class Base
+    # This class is responsible for taking a project and a list of ApiVersions (aka raw versions mapped
+    # to a common set of attributes), and creating Versions for the Project in the database, in a single
+    # query. Because we're using +upsert_all()+, Versions that already exist will just be updated.
+    # This single query avoids all the N+1s involved in a regular batch insert of Versions that we did in the past.
+    class BulkVersionUpdater
+      def initialize(db_project:, api_versions:, repository_source_name:)
+        @db_project = db_project
+        @api_versions = api_versions
+        @repository_source_name = repository_source_name
+      end
+
+      def run!
+        return if @api_versions.empty?
+
+        # create synthetic versions without saving them, so we can get their attributes
+        attrs = @api_versions
+          .map { |api_version| Version.new(api_version.to_version_model_attributes.merge(project: @db_project)) }
+          .each do |v|
+            # these will get merged properly in the upsert_all
+            v.created_at = v.updated_at = Time.current
+            # this value will get merged w/existing in the upsert_all query (see table tests in spec)
+            v.repository_sources = @repository_source_name if @repository_source_name
+            # from Version#before_save
+            v.update_spdx_expression
+            # upsert_all doesn't do validation, so ensure they're valid here.
+            v.validate!
```

Comment:
What happens if there are two identical versions for the same project id in @api_versions? +/- differences in data? Or if two of these were running for the same project at the same time?

Reply:
that should raise a unique index error in Bugsnag, which'll be desirable because we'll want to know what kind of weird data we're getting (I personally haven't ever seen a duplicate version coming back from an API, FWIW).